src/jsifier.mjs

```diff
@@ -319,39 +319,40 @@ function(${args}) {
 
     const proxyingMode = LibraryManager.library[symbol + '__proxy'];
     if (proxyingMode) {
-      if (proxyingMode !== 'sync' && proxyingMode !== 'async' && proxyingMode !== 'none') {
-        throw new Error(`Invalid proxyingMode ${symbol}__proxy: '${proxyingMode}' specified!`);
+      const possibleProxyingModes = ['sync', 'async', 'none', 'abort', 'abort_debug'];
+      if (!possibleProxyingModes.includes(proxyingMode)) {
+        throw new Error(`Invalid proxyingMode ${symbol}__proxy: '${proxyingMode}' specified! Possible modes: ${possibleProxyingModes.join(',')}`);
       }
       if (SHARED_MEMORY) {
-        const sync = proxyingMode === 'sync';
-        if (PTHREADS) {
-          snippet = modifyJSFunction(snippet, (args, body, async_, oneliner) => {
-            if (oneliner) {
-              body = `return ${body}`;
+        snippet = modifyJSFunction(snippet, (args, body, async_, oneliner) => {
+          if (oneliner) {
+            body = `return ${body}`;
+          }
+          const rtnType = sig && sig.length ? sig[0] : null;
+          const proxyFunc =
+            MEMORY64 && rtnType == 'p' ? 'proxyToMainThreadPtr' : 'proxyToMainThread';
+          var prefix = '';
+
+          if (proxyingMode == 'sync' || proxyingMode == 'async') {
+            if (PTHREADS) {
+              deps.push('$' + proxyFunc);
+              prefix = `if (ENVIRONMENT_IS_PTHREAD) return ${proxyFunc}(${proxiedFunctionTable.length}, 0, ${+(proxyingMode === 'sync')}${args ? ', ' : ''}${args});`;
             }
-            const rtnType = sig && sig.length ? sig[0] : null;
-            const proxyFunc =
-              MEMORY64 && rtnType == 'p' ? 'proxyToMainThreadPtr' : 'proxyToMainThread';
-            deps.push('$' + proxyFunc);
-            return `
-function(${args}) {
-if (ENVIRONMENT_IS_PTHREAD)
-  return ${proxyFunc}(${proxiedFunctionTable.length}, 0, ${+sync}${args ? ', ' : ''}${args});
+            if (WASM_WORKERS && ASSERTIONS) {
+              prefix += `\nassert(!ENVIRONMENT_IS_WASM_WORKER, "Attempted to call proxied function '${mangled}' in a Wasm Worker, but in Wasm Worker enabled builds, proxied function architecture is not available!");`;
+            }
+          } else if (proxyingMode == 'abort' || (proxyingMode == 'abort_debug' && ASSERTIONS)) {
+            const insideWorker = (PTHREADS && WASM_WORKERS)
+              ? '(ENVIRONMENT_IS_PTHREAD || ENVIRONMENT_IS_WASM_WORKER)'
+              : (PTHREADS ? 'ENVIRONMENT_IS_PTHREAD' : 'ENVIRONMENT_IS_WASM_WORKER');
+            prefix = `assert(!${insideWorker}, "Attempted to call function '${mangled}' inside a pthread/Wasm Worker, but this function has been declared to only be callable from the main browser thread");`;
```

Comment:
How about "but this function can only be called from the main thread"?

Using the term `main browser thread` I a little misleading because emscripten application can be run on workers where there is no main browser thread involved.  We sometimes call this "main application thread", but I think just "main thread" here is clear enough.

Reply:
The proxying modes `__proxy: 'sync'` and `__proxy: 'async'` definitely mean to run the JS code in the main browser thread, and not in the thread that executes 'main'.

I.e. in PROXY_TO_PTHREAD builds, `__proxy: 'sync'` and `__proxy: 'async'` run the proxied JS code in the main browser thread, which is not the same thread as the main application thread (the main runtime thread)

The earlier PROXY_TO_WORKER build mode (which I presume is what you mean by "can be run on workers where there is no main browser thread involved") predates pthreads/SharedArrayBuffer, and does not interact with SAB, or the proxying architecture. So these proxying modes are not relevant there when pthreads are not used. So I think saying "main browser thread" here is the correct thing to do?

Reply:
> How about "but this function can only be called from the main thread"?

Changed the message into form

```
"Attempted to call function '${mangled}' inside a pthread/Wasm Worker, but by using the proxying directive "'${proxyingMode}'", this function has been declared to only be callable from the main browser thread"
```

I would like to keep the mention about the proxy declaration that prevents the call. Otherwise a developer might be surprised to mistakenly ponder "how does it know this function is not callable from a worker?"

Reply:
> The proxying modes `__proxy: 'sync'` and `__proxy: 'async'` definitely mean to run the JS code in the main browser thread, and not in the thread that executes 'main'.
> 

The thread on which the emscripten-generated code is first run I call the `main application thread`, and it might be running in worker, in which case there is no main browser thread at all in the picture.  When you write `MAIN_THREAD_EM_ASM` or `__proxy: 'sync'` its is the main application thread that runs you code.   It just happens that this is also normally the main browser thread.

Reply:
Hmm, has there been some kind of change in configuration that I am not aware of?

My expectation:

Building with `-sPROXY_TO_WORKER`: `main()` is run in a Worker thread. `__proxy` architecture is not available there.
Building with `-pthread`: `main()` is run in main browser thread. main browser thread == main application thread, `__proxy` targets main browser thread.
Building with `-pthread -sPROXY_TO_PTHREAD`: `main()` is run in a separate Worker thread, i.e. main application thread is not main browser thread. `__proxy: 'sync'` and `MAIN_THREAD_EM_ASM()` proxy to the main _browser_ thread, and not to the main _application_ thread.

In each case, proxying does target the main browser thread. Has this changed somehow?

Reply:
No nothing has changed.   I think we are just using different terminology.  

You are correct that that the main browser thread is always that target of proxying whenever we are actually running with on the main browser there.  This is because the main runtime thread and the main browser thread are the same thing in that case.

Its only when you do something like `new Worker("load_emscripten_program")` that you would have a main runtime thread that is *not* running on the main browser thread.  We do have several users that do this.

The docs for `emscripten_is_main_runtime_thread` and `emscripten_is_main_browser_thread` support this I believe: https://github.com/emscripten-core/emscripten/blob/08a147230902cb39d9706a06ca89b08b56d1bdd0/system/include/emscripten/threading.h#L49-L56

Wwith `PROXY_TO_PTHREAD`, `main()` is run on a thread that is **neither** the main runtime thread nor the main browser thread.

Reply:
(I believe it was actually you who added those two functions back in 174a08276d027a8b32ca4da7e130cdf4a6aabbe1)

Reply:
> We do have several users that do this.

Is this case a scenario where the user manually spawns a new Worker(), and inside that Worker, they load up the  Emscripten-compiled pthread-enabled program?

Such use case was never supported originally, because of the problem that in such a manual "I loaded the app into a Worker" mode, Emscripten would still need to be able to load some stuff (a lot of stuff) into the JS context of the main application thread, in order for any kind of general proxying to work. And Emscripten runtime JS libraries were not developed with the impression that the target thread that proxying occurs to, would not be the main browser thread.

For example, if I start searching for `__proxy: 'sync'` in /src/, there are functions that assume that they are always the main browser thread:

```
emscripten_set_window_title(): document.title only exists in main browser thread
emscripten_get_screen_size(): screen.width / screen.height exist only in main browser thread
emscripten_hide_mouse(): document.styleSheets only in main browser thread
emscripten_set_canvas_size: Browser.setCanvasSize() works only in main browser thread
emscripten_get_canvas_size: Module['canvas'] only works in main browser thread
emscripten_create_worker: requires Subworkers support in order to work (https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers#spawning_subworkers)

library_html5.js: everything here that has __proxy: 'sync' expects that proxy target lands to the main browser thread
library_egl.js, library_glut.js: everything with __proxy: 'sync' expects that proxy target is the main browser thread
library_openal.js, library_html5_webgl.js, library_sdl.js: likewise
```
that is why I added the whole PROXY_TO_PTHREAD mode so that users would not need to figure out how to manually launch Emscripten content main() in a Worker - they could reuse the system-provided mechanism for that.

Although now that I think about it, iirc subworkers are supported widely, so that is no longer a problem (in 2015 Chrome did not support subworkers)

So maybe those users that do go and launch their pthread builds in a Worker are careful to not use any of the above APIs, and just instead do generic computation there? I.e. those people are not using Canvases or library_browser.js or library_html5.js or any other DOM accessing things that Emscripten provides?

Reply:
> (I believe it was actually you who added those two functions back in [174a082](https://github.com/emscripten-core/emscripten/commit/174a08276d027a8b32ca4da7e130cdf4a6aabbe1))

Yeah, I added those, but that is not for supporting the "user manually launches a pthread-enabled build inside their manually created new Worker()".

Reply:
> Is this case a scenario where the user manually spawns a new Worker(), and inside that Worker, they load up the Emscripten-compiled pthread-enabled program?
> 
> Such use case was never supported originally, because of the problem that in such a manual "I loaded the app into a Worker" mode, Emscripten would still need to be able to load some stuff (a lot of stuff) into the JS context of the main application thread, in order for any kind of general proxying to work. And Emscripten runtime JS libraries were not developed with the impression that the target thread that proxying occurs to, would not be the main browser thread.

Yes exactly.   It turns out that this configuration is supported, at least for some programs.  Obviously such programs cannot use DOM or anything like that, but I believe we do at least some testing of this configuration.

Reply:
> Yeah, I added those, but that is not for supporting the "user manually launches a pthread-enabled build inside their manually created new Worker()".

Can you explain what you intended the difference between those two functions to be then?   I seems that unless we support that configuration then then the main runtime thread and the main browser thread would always be the same?   Unless I'm missing something?

Reply:
Very originally those functions intended to be a generic set to enable common handling of each of single-threaded, -pthread, and PROXY_TO_PTHREAD and PROXY_TO_WORKER build modes, so that users would be able to do multiple variants of builds from the same codebase and we'd be able to write system library code that would be aware of each of those modes.

So, for example, in a PROXY_TO_PTHREAD build, it would be expected to happen that `emscripten_is_main_runtime_thread()` == `emscripten_is_main_browser_thread()`. But then if one would make a `PROXY_TO_WORKER` build of the same, then those functions would change to return the different respective things. So that way one could have one codebase that was ready to mutate to any of those build modes.

I.e. I did not originally intend for pthreads enabled builds to ever be loaded while inside a Worker(), since the __proxy: 'sync' semantics would proxy to the hosting Worker instead of going to the main browser thread, and we have never had any facilities to do proxying to the main browser thread in such am ode.

Although now that I think about it, yeah, it is probably the case that if user avoids all those library_x.js files that I listed above (that suffer from this problem), things will probably be pretty safe for them, even if they manually load pthread-enabled builds. And one can build "reasonable" use cases in the absence of relying on those library_x.js files.